api/src/main/java/org/infinispan/api/async/Caches.java

```diff
@@ -0,0 +1,27 @@
+package org.infinispan.api.async;
+
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+
+import org.infinispan.api.configuration.CacheConfiguration;
+
+/**
+ * @since 13.0
+ **/
+public interface Caches {
+   <K, V> CompletionStage<Cache<K, V>> create(String name, CacheConfiguration cacheConfiguration);
+
+   <K, V> CompletionStage<Cache<K, V>> create(String name, String template);
+
+   <K, V> CompletionStage<Cache<K, V>> get(String name);
```

Comment:
I propose changing the behavior of how caches are started. I personally feel that a Cache once defined should automatically start and only once the cache has fully started that it be available to the `get` method. This way this method can return immediately and instead the user would get a `CacheNotStarted` or `CacheNotDefined` error etc. from this method if the cache isn't up yet. I think this would help simplify user code in many cases as they don't have an extra layer of `CompletionStage` to worry about.

Reply:
:+1:, we already automatically start the caches that are defined in infinispan.xml or via the administration API.

The `Caches` interface doesn't have `DefaultCacheManager.defineConfiguration()`, so the only way `get()` could create a cache is if there is a matching wildcard template. We could add a `create()` override  for that job, although if the application is creating the cache programmatically it's probably safer to specify the template name.

The other case when `getCache()` now blocks is when a node creates a cache cluster-wide and then uses the cluster executor to run a task using that cache on every node. I think I would rather specify that the stage from `createCache()` completes only when the cache has been created everywhere.